// n8n-nodes-yepcode/dist/credentials/YepCodeApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YepCodeApi = void 0;
const transport_1 = require("../nodes/YepCode/transport");
class YepCodeApi {
    constructor() {
        this.name = 'yepCodeApi';
        this.displayName = 'YepCode API';
        this.documentationUrl = 'https://yepcode.io/docs';
        this.properties = [
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                description: "Find your API Token in the YepCode dashboard under the 'Settings -> API Credentials' section.",
                default: '',
                typeOptions: {
                    password: true,
                },
            },
            {
                displayName: 'API Host',
                name: 'apiHost',
                type: 'string',
                default: transport_1.DEFAULT_API_HOST,
                description: "The YepCode API host to use (default: https://cloud.yepcode.io). It can be customized for YepCode On-Premise instances.",
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'x-api-token': '={{$credentials.apiToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials?.apiHost}}',
                url: '/run/whoami',
            },
        };
    }
}
exports.YepCodeApi = YepCodeApi;
